js_modules/dagster-ui/packages/ui-core/src/runs/useRunsForTimeline.tsx

```diff
@@ -23,52 +25,151 @@ import {repoAddressAsHumanString} from '../workspace/repoAddressAsString';
+      buckets.push([time, Math.min(end, time + BUCKET_SIZE)] as const);
+    }
+    return buckets;
+  }, [start, end]);
+
+  const client = useApolloClient();
+
+  const [terminatedRunsQueryData, setTerminatedRunsData] = useState<{
+    data: RunTimelineFragment[] | undefined;
+    loading: boolean;
+    error: any;
+    called: boolean;
+  }>({
+    data: undefined,
+    loading: true,
+    error: undefined,
+    called: false,
   });
-
-  const terminatedRunsQueryData = useQuery<
-    TerimatedRunTimelineQuery,
-    TerimatedRunTimelineQueryVariables
-  >(TERMINATED_RUN_TIMELINE_QUERY, {
-    notifyOnNetworkStatusChange: true,
-    // With a very large number of runs, operating on the Apollo cache is too expensive and
-    // can block the main thread. This data has to be up-to-the-second fresh anyway, so just
-    // skip the cache entirely.
-    fetchPolicy: 'no-cache',
-    variables: {
-      terminatedFilter: {
-        ...runsFilter,
-        statuses: Array.from(doneStatuses),
-        createdBefore: endSec,
-        updatedAfter: startSec,
-      },
-    },
+  const [OngoingRunsQueryData, setOngoingRunsData] = useState<{
+    data: RunTimelineFragment[] | undefined;
+    loading: boolean;
+    error: any;
+    called: boolean;
+  }>({
+    data: undefined,
+    loading: true,
+    error: undefined,
+    called: false,
   });
+  const {data: OngoingRunsData, loading: loadingOngoingRunsData} = OngoingRunsQueryData;
+  const {data: terminatedRunsData, loading: loadingTerminatedRunsData} = terminatedRunsQueryData;
+
+  const fetchTerminatedRunsQueryData = useCallback(async () => {
+    return await fetchPaginatedBucketData({
+      buckets,
+      setQueryData: setTerminatedRunsData,
+      async fetchData(bucket, cursor: string | undefined) {
+        const {data} = await client.query<
+          TerminatedRunTimelineQuery,
+          TerminatedRunTimelineQueryVariables
+        >({
+          query: TERMINATED_RUN_TIMELINE_QUERY,
+          notifyOnNetworkStatusChange: true,
+          fetchPolicy: 'no-cache',
+          variables: {
+            terminatedFilter: {
+              ...runsFilter,
+              statuses: Array.from(doneStatuses),
+              createdBefore: bucket[1] / 1000,
+              updatedAfter: bucket[0] / 1000,
+            },
+            cursor,
+            limit: BATCH_LIMIT,
+          },
+        });
+
+        if (data.terminated.__typename !== 'Runs') {
+          return {
+            data: [],
+            cursor: undefined,
+            hasMore: false,
+            error: data.terminated,
+          };
+        }
+        const runs = data.terminated.results;
+        const hasMoreData = runs.length === BATCH_LIMIT;
+        const nextCursor = hasMoreData ? runs[runs.length - 1]!.id : undefined;
+        return {
+          data: runs,
+          cursor: nextCursor,
+          hasMore: hasMoreData,
+          error: undefined,
+        };
```

Comment:
In the future this `fetchData` function could check a cache first for the particular bucket